### PR TITLE
Mach-O File Format Parsing

### LINKED ISSUE #1561: [FR]: Add Mach-O parsing functionality
### What is the feature you are proposing?

Mach-O file parsing is currently not available in Turbinia.
Large enterprises with a considerable fleet of MacOS devices would benefit of having this feature available.

### What would this feature improve or what problem would it solve?

Having Mach-O parsing capabilities would significantly enhance the reach of available file formats that Turbinia can parse.

### What alternatives have you considered?

_No response_

### PR DESCRIPTION
**Pull request by daschwanden**
### Description of the change

This PR proposes adding a Mach-O file format parser to Turbinia.

### Applicable issues

- fixes #1561

### Additional information

Mach-O parsing is critical to large enterprises with large fleets of MacOS devices.
Having such functionality available could greatly enhance the reach of Turbinia.

### Checklist

- [X] All tests were successful.
- [X] Unit tests added.
- [X] Documentation updated.

### CODE REVIEW

File: turbinia/workers/analysis/macho.py
Comment (aarontp) [inline turbinia/workers/analysis/macho.py:324]: Minor nit:  I understand that these log lines are indented because it's in a loop and there will be multiple of them, but there are a few other log lines below in the same section that have varying levels if indentation.  Is there a reason they're not at a single indent similar to the above line?

File: turbinia/workers/analysis/macho.py
Comment (aarontp) [inline turbinia/workers/analysis/macho.py:414]: There's a few log lines with various formats of prefix/postfix characters (dashes, asterisks, and equals signs) and capitalization.  This is going to a normal log file, so usually we don't have much other markup data here, but I can see how it might be useful in this case when logging lots of structured data.   Can you send me the output of what this looks like?  I wondering if we can make the formatting a bit more consistent.

**Review by aarontp — COMMENTED** [review]
Overall it LGTM, mostly just some minor nits and some questions.

File: pyproject.toml
Comment (daschwanden) [reply pyproject.toml:73]: From a functionality point of view the newly added libraries (asn1crypto, lief, pyssdeep and py-tlsh) are only required by the worker. However, for reasons unclear to me the server and api_server also want the dependency fullfilled. Do you know why both the server and api_server would be interested in worker dependencies?

File: docker/local/docker-compose-dev.yml
Comment (daschwanden) [reply docker/local/docker-compose-dev.yml:1]: It is a copy of an existing docker compose file as I did not want to mess with the existing setup.

It has been used for development using the 'docker compose watch' feature (see lines 68 - 72, more details on watch here: https://docs.docker.com/compose/how-tos/file-watch/).

docker compose watch significantly speeds up the inner dev loop as it constantly syncs any code changes to the running container.

Not required that we keep it. Can delete it in case you feel so.
However, in case you reckon it is worthwhile keeping it around then we could add some documentation similar to what we added to the GRR docs (see [Section 4 here](https://docs.docker.com/compose/how-tos/file-watch/) for an example).

File: turbinia/workers/analysis/macho.py
Comment (daschwanden) [reply turbinia/workers/analysis/macho.py:216]: Totally, will remove the commented out log lines and the `result` where not needed.

File: turbinia/workers/analysis/macho.py
Comment (daschwanden) [reply turbinia/workers/analysis/macho.py:491]: Good catch. This `try` should be removed now as if it fails it should actually fail. In case the binary has a `code_signature` section the parsing has to work or it deserves an exception to be thrown.

File: turbinia/workers/analysis/macho.py
Comment (daschwanden) [reply turbinia/workers/analysis/macho.py:324]: This output was for development only. Will remove as it is no longer required.

File: turbinia/workers/analysis/macho.py
Comment (daschwanden) [reply turbinia/workers/analysis/macho.py:414]: Good point. I will clean up and unify the output to match what we do in the rest of the code currently and remove  logging lines that are no longer needed.

File: turbinia/workers/analysis/macho.py
Comment (daschwanden) [reply turbinia/workers/analysis/macho.py:580]: Don't think we should come across this case but reckon leaving your suggested log line would make sense just in case. Will update accordingly.

File: turbinia/workers/analysis/macho.py
Comment (daschwanden) [reply turbinia/workers/analysis/macho.py:582]: Good point. Will remove it.

File: turbinia/workers/analysis/macho.py
Comment (daschwanden) [reply turbinia/workers/analysis/macho.py:565]: There is a single ('wrapper') fat binary that contains one or more platform architecture specific binaries (arm, x86). Hence, there can only be one binary for one architecture each. This is Apple's way to make sure they can support different platforms with a single 'wrapper binary'.

File: turbinia/workers/analysis/macho.py
Comment (daschwanden) [reply turbinia/workers/analysis/macho.py:481]: Plain vanilla json.dumps() doesn't support custom classes.
See https://stackoverflow.com/questions/3768895/how-to-make-a-class-json-serializable
Added a comment in code to point the reader to the reasoning.

File: turbinia/workers/analysis/macho.py
Comment (daschwanden) [reply turbinia/workers/analysis/macho.py]: Yep, that is much nicer :-)

File: pyproject.toml
Comment (jleaniz) [inline pyproject.toml:80]: Drive by comment but i don't understand why you need to add these dependencies to the API server, or the server. These should only be required for the worker. Definitely do not want to add plaso as a dependency to the api server container here.

**Review by jleaniz — CHANGES_REQUESTED** [review]
Drive by comment

File: pyproject.toml
Comment (jleaniz) [reply pyproject.toml:73]: This shouldn't be the case. Do you have any indication of where you run into this? Is it during container build or runtime?

File: pyproject.toml
Comment (daschwanden) [reply pyproject.toml:80]: Totally agree. This has surprised me too.
Have emailed you the docker logs in for server and api-server in case we do not include them.
Let's have a quick chat and hope we might be able to identify what the reason might be.

File: pyproject.toml
Comment (daschwanden) [reply pyproject.toml:73]: It is at run time that server and api-server complain.
Have emailed you the output.
Let's have a quick chat and see if we can get to the bottom of this.

File: pyproject.toml
Comment (daschwanden) [reply pyproject.toml]: Thanks heaps Juan for your help via email.
Have implemented your suggestion.
Hope it addresses your feedback.

**Review by jleaniz — APPROVED** [review]
LGTM regarding the changes i requested.

File: docker/local/docker-compose-dev.yml
Comment (aarontp) [reply docker/local/docker-compose-dev.yml:1]: @hacktobeer Do you have any thoughts on this one?  Do you think it would be beneficial to keep this one, or possibly integrate the relevant changes to one of the other existing dev compose files?

**Review by aarontp — APPROVED** [review]
Changes LGTM.  Will wait to merge until after the question about whether to include the docker file is resolved.

**Comment by aarontp** [comment]
Hi @daschwanden,  I think this is ready to merge after the merge conflicts have been resolved.   It's probably easiest just to re-run `poetry lock` to update that file.  Thanks!

**Comment by daschwanden** [comment]
> Hi @daschwanden, I think this is ready to merge after the merge conflicts have been resolved. It's probably easiest just to re-run `poetry lock` to update that file. Thanks!

Hi @aarontp , thanks heaps for resolving the merge conflicts. I am wondering whether the best way to resolve the remaining `poetry.lock` conflict would be to run `poetry lock` on both the upstream and the fork feature branch and then commit both. What do you think?

**Comment by aarontp** [comment]
> Hi @aarontp , thanks heaps for resolving the merge conflicts. I am wondering whether the best way to resolve the remaining `poetry.lock` conflict would be to run `poetry lock` on both the upstream and the fork feature branch and then commit both. What do you think?

I mentioned this out of band as well, but I think it's fine to update all the dependencies in this PR (without a separate one) assuming all of the tests pass, etc.  You might need to revert just the poetry.lock file in your PR and then re-run `poetry lock`.

**Comment by daschwanden** [comment]
> > Hi @aarontp , thanks heaps for resolving the merge conflicts. I am wondering whether the best way to resolve the remaining `poetry.lock` conflict would be to run `poetry lock` on both the upstream and the fork feature branch and then commit both. What do you think?
> 
> I mentioned this out of band as well, but I think it's fine to update all the dependencies in this PR (without a separate one) assuming all of the tests pass, etc. You might need to revert just the poetry.lock file in your PR and then re-run `poetry lock`.

Hi @aarontp , managed to resolve the conflict using the following excellent [blog post](https://www.peterbe.com/plog/how-to-resolve-a-git-conflict-in-poetry.lockl). Noteing it here so in case we or someone run into this issue in the future again.

**Pull request merged**
